### PR TITLE
Invalid `?` suggestion on mismatched `Ok(T)`

### LINKED ISSUE #116967: Invalid `?` suggestion on mismatched `Ok(T)`
### Code

```Rust
pub fn foo() -> Result<String, ()> {
    let out: Result<(), ()> = Ok(());
    out
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/lib.rs:3:5
  |
1 | pub fn foo() -> Result<String, ()> {
  |                 ------------------ expected `Result<String, ()>` because of return type
2 |     let out: Result<(), ()> = Ok(());
3 |     out
  |     ^^^ expected `Result<String, ()>`, found `Result<(), ()>`
  |
  = note: expected enum `Result<String, _>`
             found enum `Result<(), _>`
help: use `?` to coerce and return an appropriate `Err`, and wrap the resulting value in `Ok` so the expression remains of type `Result`
  |
3 |     Ok(out?)
  |     +++   ++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to previous error
```

### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/lib.rs:3:5
  |
1 | pub fn foo() -> Result<String, ()> {
  |                 ------------------ expected `Result<String, ()>` because of return type
2 |     let out: Result<(), ()> = Ok(());
3 |     out
  |     ^^^ expected `Result<String, ()>`, found `Result<(), ()>`
  |
  = note: expected enum `Result<String, _>`
             found enum `Result<(), _>`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to previous error
```

### Rationale and extra context

The suggestion is invalid

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by eopb**
fixes: #116967

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 1]
```diff
@@ -983,6 +983,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         } else {
             return false;
         }
+        let e_ok = args_e.type_at(0);
+        let f_ok = args_f.type_at(0);
+        if !self.infcx.can_eq(self.param_env, f_ok, e_ok) {
+            return false;
```
Comment (Nilstrieb) [inline compiler/rustc_hir_typeck/src/demand.rs:989]: unrelated to your change but I feel like all of this would read way nicer as one big let chain.

Comment (eopb) [reply compiler/rustc_hir_typeck/src/demand.rs:989]: Good point, I'll consider doing a little refactor of this function after this PR is merged

Comment (eopb) [reply compiler/rustc_hir_typeck/src/demand.rs]: How's this? https://github.com/rust-lang/rust/pull/116968/commits/24cdb27e2842e5abab2875ce29c365ac5503773e

**Pull request merged**